vllm/attention/backends/xformers.py

```diff
@@ -384,9 +392,166 @@ def _get_seq_len_block_table_args(
         raise AttributeError(f"Invalid attention type {str(attn_type)}")
 
 
-class XFormersMetadataBuilder(CommonMetadataBuilder[XFormersMetadata]):
+class XFormersMetadataBuilder(AttentionMetadataBuilder[XFormersMetadata]):
```

Comment:
Why we need to re-implement the metadata builder? What's the difference between this implementation and the common one?

Reply:
It has #7018 and

```py
            if prefix_cache_hit:
                # NOTE(woosuk): For xformers, the block table should
                # include the entries for the incoming prefill tokens.
                block_table = block_tables[seq_id]
            elif ((chunked_prefill_enabled or not is_prompt)
                  and block_tables is not None):
                if curr_sliding_window_block == 0:
                    block_table = block_tables[seq_id]
                else:
                    block_table = block_tables[seq_id][
                        -curr_sliding_window_block:]
```

from flash-attn. However, it is based on the common one, not just copied from flash-attn. Theoretically, I could modify the common, but that might affect other attention backends.

Reply:
In fact, it works even without #7018, so I'll remove the changes from #7018.

Reply:
Can you try to modify the common one? We should avoid to introduce this divergence as possible.

Reply:
Done